cpp/independent_set/IndependentSet.cpp

```diff
@@ -0,0 +1,318 @@
+    for (auto v : nodes) {
+        graph->restoreNode(v);
+    }
+    for (auto e : edges) {
+        graph->addEdge(e.u, e.v);
+    }
+}
+
+std::vector<NetworKit::node> IndependentSet::runIndependentSetDegree2() const {
+    NetworKit::Graph graphDeg2(*graph);
+    std::vector<NetworKit::node> independentSet;
+
+    auto solvePath = [&] (NetworKit::node u) { // u -> v -> w
+        while (graphDeg2.hasNode(u) && graphDeg2.degree(u) <= 1) {
+            if (graphDeg2.degree(u) == 0) {
+                independentSet.push_back(u);
+                graphDeg2.removeNode(u);
+                break;
+            }
+
+            independentSet.push_back(u);
+            NetworKit::node v = graphDeg2.getIthNeighbor(u, 0);
+            graphDeg2.removeNode(u);
+
+            if (graphDeg2.degree(v) == 1) {
+                NetworKit::node w = graphDeg2.getIthNeighbor(v, 0);
+                u = w;
+            }
+            graphDeg2.removeNode(v);
+        }
+    };
+
+    graphDeg2.forNodes([&](NetworKit::node u) {
+        solvePath(u);
+    });
+
+    graphDeg2.forNodes([&](NetworKit::node u) { // t <- u -> v -> w
+        if (graphDeg2.degree(u) == 2) {
+            independentSet.push_back(u);
+            NetworKit::node t = graphDeg2.getIthNeighbor(u, 0);
+            graphDeg2.removeNode(t);
+            NetworKit::node v = graphDeg2.getIthNeighbor(u, 0);
+            graphDeg2.removeNode(u);
+
+            if (graphDeg2.degree(v) > 0) {
+                NetworKit::node w = graphDeg2.getIthNeighbor(v, 0);
+                graphDeg2.removeNode(v);
+                solvePath(w);
+            }
+            else {
+                graphDeg2.removeNode(v);
+            }            
+        }
+    });
+    return independentSet;
+}
+
+IndependentSet::IndependentSet(const NetworKit::Graph &graph) : 
+        graph(std::make_optional(graph)) { }
+
+bool IndependentSet::edgeComparator(const NetworKit::Edge& a, const NetworKit::Edge& b) {
+    return a.u < b.u || (a.u == b.u && a.v < b.v);
+}
+
+const std::map<NetworKit::node, bool>& IndependentSet::getIndependentSet() const {
+    assureFinished();
+    return independentSet;
+}
+
+void BruteForceIndependentSet::run() {
+    if (graph->isEmpty()) {
+        return;
+    }
+
+    std::vector<bool> testSet;
+    graph->forNodes([&](NetworKit::node v) {
+        testSet.push_back(false);
+        independentSet[v] = false;
+    });
+
+    int best = 0;
+    unsigned long long max = 1;
+    max <<= graph->numberOfNodes();
+
+    for (unsigned long long binary = 1; binary != max; ++binary) {
+        std::bitset<8 * sizeof(unsigned long long)> testSet(binary);
+        size_t testSetSize = testSet.count();
+        if (testSetSize <= best) {
+            continue;
+        }
+
+        bool illegalEdge = false;
+        graph->forEdges([&](NetworKit::node u, NetworKit::node v) {
+            if(testSet[u] && testSet[v]) {
+                illegalEdge = true;
+                return;
```

Comment:
This return does not improve the efficiency, as we traverse through all the edges. If you want a speedup, you need to rewrite it as "for (auto e = graph->edgeRange().begin(); e != graph->edgeRange().end(); e++) {...}